### PR TITLE
Complete redesign of the resource registry

### PR DESCRIPTION
**Pull request by edsko**

### CODE REVIEW

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 1]
```diff
@@ -166,7 +166,7 @@ validateChainDb
   -> Bool -- Verbose
   -> IO ()
 validateChainDb dbDir cfg onlyImmDB verbose =
-  ResourceRegistry.with $ \registry -> give protocolMagicId $ give epochSlots $
+  ResourceRegistry.withRegistry $ \registry -> give protocolMagicId $ give epochSlots $
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Remove the qualifier, now we have `Registry` 3x on one line. It's almost as bad as Java 😀

File: ouroboros-consensus/test-util/Test/Util/SOP.hs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE ScopedTypeVariables #-}
```
Comment (mrBliss) [inline ouroboros-consensus/test-util/Test/Util/SOP.hs]: You know that there's somebody who would want you to make this change in a separate commit 🙂

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs [hunk 3]
```diff
@@ -62,55 +63,51 @@ data BlockchainTime m = BlockchainTime {
...
+
+      -- | Spawn a thread to run an action when reaching the specified slot
+      --
+      -- The thread will be linked to the registry in which the 'BlockchainTime'
+      -- itself was created.
+      --
+      -- Unlike the thread created by 'onSlotChange_', this thread will
+      -- terminate as soon as the action does.
+      --
+      -- Use sites should call 'onSlot_' rather than 'onSlot_'.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: The first `onSlot_` should be `onSlot`

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 4]
```diff
@@ -96,7 +96,7 @@ run
   -> IO ()
 run tracers chainDbTracer rna dbPath pInfo
     customiseChainDbArgs customiseNodeArgs onNodeKernel =
-    ResourceRegistry.with $ \registry -> do
+    ResourceRegistry.withRegistry $ \registry -> do
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I would remove the qualifier

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 5]
```diff
@@ -273,22 +273,26 @@ initNetwork
   -> IO ()
 initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
     -- serve local clients (including tx submission)
-    localServer <- forkLinked registry runLocalServer
+    localServer <- forkThread registry runLocalServer
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Would it sense to add `forkLinkedThread`?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 6]
```diff
@@ -273,22 +273,26 @@ initNetwork
...
     -- dns subscription managers
-    dnsSubscriptions <- forM rnaDnsProducers $ \dnsProducer ->
-      forkLinked registry $ runDnsSubscriptionWorker connTable dnsProducer
+    dnsSubscriptions <- forM rnaDnsProducers $ \dnsProducer -> do
+       worker <- forkThread registry $
+                   runDnsSubscriptionWorker connTable dnsProducer
+       linkToRegistry worker
+       return worker
 
     let asyncs = localServer : peerServer : ipSubscriptions : dnsSubscriptions
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: `threads` would be more accurate than `asyncs`

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs [hunk 7]
```diff
@@ -82,10 +82,15 @@ launchBgTasks cdb@CDB{..} = do
       copyToImmDBRunner cdb gcSchedule
     lgrSnapshotThread <- launch $
       updateLedgerSnapshotsRunner cdb
+    -- TODO: Do we still need this cdbBgThreads, if we have the registry?
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs]: Hmm, I think the only reason we have `cdbBgThreads` is to be able to shut down the background threads before we close the ImmutableDB and the VolatileDB, so that no background thread tries to access them after their closure, which would result in a `ClosedDBError`.

Since the ImmutableDB and the VolatileDB are not managed by the `ResourceRegistry` (maybe they should be?!), these actions do not take place in the right order, I believe.

File: ouroboros-consensus/tools/db-convert/Main.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/tools/db-convert/Main.hs]: Yeah, fair enough. Just wanted to make minimal changes to stuff not directly in consensus. Will change.

File: ouroboros-consensus/test-util/Test/Util/SOP.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/test-util/Test/Util/SOP.hs]: Thankfully that someone is on holidays :grin:

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 5 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: It seemed unnecessary, but since both you and @nfrisby prefer it, I'll add it :)

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: I'll remove the `forkThreadSelf`, I thought we'd need it in tests but turns out it doesn't help. And it avoids requiring a `forkLinkedThreadSelf` variant then too.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs (hunk 7 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs]: I'll leave it for now. It kinda feels to me that this `MVar` is a registry in disguise, and so maybe we should just use one, but it's not essential to this PR.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 8]
```diff
@@ -1,324 +1,763 @@
...
+-- > withRegistry $ \registry ->
+-- >   r <- allocate registry allocateResource releaseResource
+-- >   fork $ .. use r ..
+--
+-- It is easy to see that this code is problematic: we allocate a resource @r@,
+-- then spawn a thread that uses @r@, and finally leave the scope of
+-- 'withRegistry', thereby deallocating @r@ -- leaving the thread to run with
+-- a now deallocated resource.
+--
+-- It is /only/ safe for threads to use a given registry if the lifetime of
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: Yup, all sounds good to me. Made the changes.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 9]
```diff
@@ -1,324 +1,763 @@
...
+--
+-- For this reason, the registry's 'linkToRegistry' combinator does not link the
+-- specified to the thread calling 'linkToRegistry', but rather to the thread
+-- that created the registry. After all, the lifetime of threads spawned with
+-- 'forkThread' can certainly exceed the lifetime of their parent threads, but
+-- the lifetime of /all/ threads spawned using the registry will be limited by
+-- the scope of that registry, and hence the lifetime of the thread that created
+-- it. So, when we call 'linkToRegistry', the exception will be thrown the
+-- thread that created the registry, which (if not caught) will cause that that
+-- to exit the scope of 'withRegistry', thereby terminating the entire hierarchy
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: Addressing the first point: renamed `newRegistry` to `unsafeNewRegistry`. Also added a section "Combinators primarily for testing" in the exports, and move `unsafeNewRegistry`, `closeRegistry` as well as `countResources` to that section.

[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: As regards terminology, I'd rather we assume less rather than more, and be explicit where needed. I've added a note:

```
-- Note on terminology: when thread A forks thread B, we will say that thread A
-- is the " parent " and thread B is the " child ". No further relationship
-- between the two threads is implied by this terminology. In particular, note
-- that the child may outlive the parent. We will use "fork" and "spawn"
-- interchangeably.
```

I agree that "hierarchy" is indeed misleading, and I've rephrased the text to avoid it (details below).

[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: I've weakened the "chaining" paragraph to say:

```
-- Linking a parent to a child is however of limited value if the lifetime of
-- the child is not limited by the lifetime of the parent. For example, if A
-- does
--
-- > threadB <- async $ ..
-- > link threadB
--
-- and A terminates before B does, any exception thrown by B might be send to a
-- thread that no longer exists. This is particularly problematic when we start
-- chaining threads: if A spawns-and-links-to B which spawns-and-links-to C, and
-- C throws an exception, perhaps the intention is that this gets rethrown to B,
-- and then rethrown to A, terminating all three threads; however, if B has
-- terminated before the exception is thrown, C will throw the exception to a
-- non-existent thread and A is never notified.
```

[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: Avoiding another use of hierarchy:

```
-- thread that created the registry, which (if not caught) will cause that that
-- to exit the scope of 'withRegistry', thereby terminating all threads in that
-- registry.
```

[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: The last use of "hierachy" is justified:

```
-- This /also/ includes nested calls to 'withRegistry'. Since the lifetime of
-- such a registry (and all resources within) is tied to the thread calling
-- 'withRegistry', which itself is tied to the "parent registry" in which it was
-- created, this creates a hierarchy of registries.
```

[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: > Is that too thin of a scenario to accommodate without a concrete example to back it up? 

No, I don't think so, I thought about it also, but I guess I wasn't specific enough in the documentation. I've added the following part to the last section:

```
-- This includes 'withAsync': after all, 'withAsync' can't result in a child
-- thread that outlives its parent, so this is safe. In particular, if a parent
-- thread wants to link to a child thread /and handle any exceptions arising
-- from the link/, it should probably use `withAsync`:
--
-- > let handleLinkException :: ExceptionInLinkedThread -> m ()
-- >     handleLinkException = ..
-- > in handle handleLinkException $
-- >      withAsync codeInChild $ \child ->
-- >        ..
--
-- instead of
--
-- > handle handleLinkException $ do  -- PROBABLY NOT CORRECT!
-- >   child <- async codeInChild
-- >   ..
--
-- where the parent may exit the scope of the exception handler before the child
-- terminates. If the lifetime of the child cannot be limited to the lifetime of
-- the parent, the child should probably be linked to the registry instead and
-- the thread that spawned the registry should handle any exceptions.
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 10]
```diff
@@ -1,324 +1,791 @@
...
+      Left closed ->
+        throwRegistryClosed context closed
+      Right key -> mask_ $ do
+        a <- alloc key
+        inserted <- updateState rr $ insertResource key (mkResource context a)
+        case inserted of
+          Left closed -> do
+            -- Despite the earlier check, it's possible that the registry got
+            -- closed after we allocated a new key but before we got a chance to
+            -- register the resource. In this case, we must deallocate the
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: Have left a comment about this in the code for now.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs [hunk 11]
```diff
@@ -1,324 +1,791 @@
...
+-- | Cancel a thread
+--
+-- This is a synchronous operation: the thread will have terminated when this
+-- function returns.
+cancelThread :: MonadAsync m => Thread m a -> m ()
+cancelThread = uninterruptibleCancel . threadAsync
+
+-- | Wait for thread to terminate and return its result.
+--
+-- If the thread throws an exception, this will rethrow that exception.
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs]: Good point. Added comment:

```
-- NOTE: If A waits on B, and B is linked to the registry, and B throws an
-- exception, then A might /either/ receive the exception thrown by B /or/
-- the 'ThreadKilled' exception thrown by the registry.
```

**Comment by edsko** [comment]
Addressed the feedback from @mrBliss and @nfrisby . Moreover, did another thorough peer review over hangout with @nfrisby which resulted in a few subtle tweaks. I think this is ready for a final review by @nfrisby and can then be merged.
